### ISSUE TITLE
TimeoutError: Promise timed out after 90000 milliseconds

### ISSUE THREAD

**Issue by Keller18306** [telegraf/telegraf#1479]
* Telegraf.js Version: 4.3.0
* Node.js Version: 14.16.1
* Operating System: win10

## Minimal Example Code Reproducing the Issue
![image](https://user-images.githubusercontent.com/23178771/123495723-86e59f80-d62d-11eb-96bb-e3615493c349.png)
![image](https://user-images.githubusercontent.com/23178771/123495682-5f8ed280-d62d-11eb-9362-8765788bf479.png)

I think this happens if you upload large files with slow internet

## Expected Behavior

The files must be uploaded to the telegram servers, sent to the user, return the array of uploaded files

## Current Behavior

The files were uploaded to the telegram servers, sent to the user. But this error returned. No file array returned. The execution of further code has been abandoned due to an error.
![image](https://user-images.githubusercontent.com/23178771/123495641-3e2de680-d62d-11eb-8e13-a570249fd930.png)

## Error Message and Logs (`export DEBUG='telegraf:*'`)
I do not have

**Comment by Keller18306** [comment]
catch in this method doesn't help
![image](https://user-images.githubusercontent.com/23178771/123496314-23a93c80-d630-11eb-968c-c46a404065b5.png)
![image](https://user-images.githubusercontent.com/23178771/123496305-15f3b700-d630-11eb-8183-cbebb6a27a11.png)

**Comment by Keller18306** [comment]
> catch in this method doesn't help
> ![image](https://user-images.githubusercontent.com/23178771/123496314-23a93c80-d630-11eb-968c-c46a404065b5.png)
> ![image](https://user-images.githubusercontent.com/23178771/123496305-15f3b700-d630-11eb-8183-cbebb6a27a11.png)

Also, after that, the bot stopped responding to my messages.

**Comment by Keller18306** [comment]
I noticed that this also happens with one audio. When I pressed multiple buttons at once.
![image](https://user-images.githubusercontent.com/23178771/123498529-26f4f600-d639-11eb-9da8-b31ef73c190a.png)
![image](https://user-images.githubusercontent.com/23178771/123498534-3116f480-d639-11eb-9158-0273faaf5793.png)
![image](https://user-images.githubusercontent.com/23178771/123498538-37a56c00-d639-11eb-88d8-77d1a6b69c39.png)
For some reason, the behavior of my bot has changed. Previously, I could press several buttons at once and he executed them in parallel. Now some kind of "queue" has appeared.

**Comment by Keller18306** [comment]
My hypothesis is correct, that this is due to the slow internet.  It turns out I had VPN turned on.

**Comment by NelsonHernandezM** [comment]
did you find how to fix it?

**Comment by ppabcd** [comment]
How to catch this error.

**Comment by Reimuna** [comment]
I'm getting the same here

**Comment by vakhamzin** [comment]
I use [ngrok](https://ngrok.com/) for local development. Can it be the reason for this issue?

**Comment by gthrm** [comment]
**telegraf** is using a [p-timeout](https://www.npmjs.com/package/p-timeout) for Promises (https://github.com/telegraf/telegraf/blob/v4/src/telegraf.ts#L232)
By default:
```javascript
const DEFAULT_OPTIONS: Telegraf.Options<Context> = {
  telegram: {},
  handlerTimeout: 90_000, // 90s in ms
  contextType: Context,
}
```
Use `export const bot = new Telegraf(BOT_TOKEN, {handlerTimeout: 9_000_000});`
Where `handlerTimeout` (for example) is 2.5 hours

**Comment by ezimnitskiy** [comment]
@gthrm Нow will this solve the problem if 1000 users simultaneously hang for 2.5 hours waiting? How to solve the problem globally so that you can just catch this exception? I get such an error on ctx.reply, when the user clicks on the inline button during the period when about 100 users simultaneously knock on the bot at the same time

**Comment by gthrm** [comment]
> @gthrm Нow will this solve the problem if 1000 users simultaneously hang for 2.5 hours waiting? How to solve the problem globally so that you can just catch this exception?

It's about "TimeoutError: Promise timed out after 90000 milliseconds". Not about catch and 1000 users. For example, for 1000 users you can create a queue. And create PR for fix situation with catch. You know?

**Comment by ezimnitskiy** [comment]
@gthrm why don't you just add a construct to catch this exception in bot.catch?

**Comment by ezimnitskiy** [comment]
Or are we going to play fingers? :)

**Comment by gthrm** [comment]
> why don't you just add

but why don't **you** just add it? 🤔🤷‍♂️

**Comment by ezimnitskiy** [comment]
@gthrm Sorry, I looked at the code, you have everything, but the error is still not caught and the bot hangs

**Comment by ezimnitskiy** [comment]
> > @gthrm Нow will this solve the problem if 1000 users simultaneously hang for 2.5 hours waiting? How to solve the problem globally so that you can just catch this exception?
> 
> It's about "TimeoutError: Promise timed out after 90000 milliseconds". Not about catch and 1000 users. For example, for 1000 users you can create a queue. And create PR for fix situation with catch. You know?

I can do it, but I would not store it in memory, I will need to use some kind of bull, and there is redis. Is that okay for you?

**Comment by snowfluke** [comment]
Interesting discussion

**Comment by KnorpelSenf** [comment]
I still don't understand from your conversation what problem you're actually trying to solve, but you are free to open a PR and explain it there.

> I will need to use some kind of bull, and there is redis

It is unlikely that a PR will be merged that makes Telegraf add bull or redis as a dependency. If you need external queuing solutions, you should look into creating a separate package that works with Telegraf and that adds support for these things.

**Comment by zb-venom** [comment]
I use 
```
bot.catch(e => {
  console.log("aaaaaaaaaaaaaaaaaaaa");
});
```

**Comment by FidelVe** [comment]
> My hypothesis is correct, that this is due to the slow internet. It turns out I had VPN turned on.

im having the same error and I also believe is because of slow internet on my side. The promises timeout because (in my case) after sending a get request my connection crashes.

**Comment by Natteke** [comment]
I dont know why,
but i have an error event with 1000ms.
```
TimeoutError: Promise timed out after 1000 milliseconds
  at Timeout._onTimeout (/home/dev1/Projects/BottleBot/node_modules/p-timeout/index.js:39:64)
  at listOnTimeout (node:internal/timers:559:17)
```
idk why, handlerTimeout option does not work.

```
const botOptions = {
  handlerTimeout: 90_000,
};
const bot = new Telegraf<BottleBotTContext>(BOT_TOKEN, botOptions);
```

in my message handler i generate an image and send it as response. It takes about 3s on my single-core host

**Comment by wojpawlik** [comment]
I added `fetch` timeout in e614a84fb59336cd39e87da2674e1e45bc792e9c, so I'm considering defaulting `handlerTimeout` to `Infinity` and deprecating it.

**Comment by MKRhere** [comment]
Timeout getting triggered maybe a sign of a larger problem; if someone has a specific case like OP where it is thrown due to large files and slow internet, I'd advice them to set `handlerTimeout` to a higher value rather than remove it from Telegraf prematurely.

**Comment by TheTinkerBasement** [comment]
I have the same problem, why can't I just catch this error?

**Comment by ppabcd** [comment]
make sure using await in every async function

**Comment by AliAryanTech** [comment]
not fixed yet

**Comment by hypernova7** [comment]
@AliAryanTech Using `Infinity` or `Number.POSITIVE_INFINITY` with `handlerTimeout` option disables the timeout according to the [p-timeout@v4.1.0 line:17](https://github.com/sindresorhus/p-timeout/blob/4f86930f75d1565927790ab70b4ac643e47007fc/index.js#L17) module code

**Comment by AliAryanTech** [comment]
> @AliAryanTech Using `Infinity` or `Number.POSITIVE_INFINITY` with `handlerTimeout` option disables the timeout according to the [p-timeout@v4.1.0 line:17](https://github.com/sindresorhus/p-timeout/blob/4f86930f75d1565927790ab70b4ac643e47007fc/index.js#L17) module code

How will you implement this with your code!? Can you show an example?

**Comment by hypernova7** [comment]
@AliAryanTech 
```javascript
const bot = new Telegraf(token, {
  handlerTimeout: Infinity // or Number.POSITIVE_INIFINITY
})
```

**Comment by AliAryanTech** [comment]
> @AliAryanTech
> 
> ```js
> const bot = new Telegraf(token, {
>   handlerTimeout: Infinity // or Number.POSITIVE_INIFINITY
> })
> ```

**I see, thanks**

**Comment by hypernova7** [comment]
@AliAryanTech please close this issue, if it has been resolved

**Comment by AliAryanTech** [comment]
> @AliAryanTech please close this issue, if it has been resolved

It's not mine
@Keller18306 opened this issue on Jun 26, 2021

**Comment by hypernova7** [comment]
@AliAryanTech Lol it's true, sorry xd

**Comment by MKRhere** [comment]
Closing as completed; this is intended behaviour and a workaround has been shared.

**Comment by AliAryanTech** [comment]
> @AliAryanTech Lol it's true, sorry xd

Sorry to bother you!
Did you made any open source bot or you know any repo which is module Base using `telegraf`

I'm having some issues so I thought I could find some examples there

**Comment by MKRhere** [comment]
@AliAryanTech Please keep the discussion on topic; don't post unrelated comments on an old issue. Raise a new issue or discussion if you need it. Meanwhile, there are some examples [here](https://github.com/feathers-studio/telegraf-docs/tree/master/examples).

**Comment by AliAryanTech** [comment]
> @AliAryanTech Please keep the discussion on topic; don't post unrelated comments on an old issue. Raise a new issue or discussion if you need it. Meanwhile, there are some examples [here](https://github.com/feathers-studio/telegraf-docs/tree/master/examples).

My bad, anyway thanks
